### DISCUSSION TITLE
405 method not allowed for http://authelia:9091/api/firstfactor

### DISCUSSION THREAD

**Discussion by jozefn** [authelia/authelia#5794] [Troubleshooting]
I have a docker network set up and running containers nginx redis authelia - the network is auth_net based on 192.168.1.0/24
nginx is 192.168.1.10
authelia is 192.168.1.11

the site to be protected is twofactor.jnorris.net - using hosts file for authentication:
192.168.1.10 twofactor.jnorris.org twofactor
192.168.1.11 authelia.jnorris.org authelia
192.168.1.12 redis.jnorris.org redis

I am using self signed certificates for this project

docker compose up -d brings everything up as I have it configured in the configuration files attached.  When I got to https://twofactor.jnorris.org I am redirected to login screen.

I enter my credentials and authelia spins a bit but then returns returns a username/password incorrect 

docker logs authelia tells a different story:

time="2023-08-05T15:17:53-06:00" level=error msg="Scheme of target URL http://authelia:9091/api/verify must be secure since cookies are only transported over a secure connection for security reasons" method=GET path=/api/verify remote_ip=192.168.1.10 stack="github.com/authelia/authelia/v4/internal/handlers/handler_verify.go:451 VerifyGET.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54       (*BridgeBuilder).Build.func1.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:16      SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.14/router.go:427                        (*Router).Handler\ngithub.com/valyala/fasthttp@v1.43.0/http.go:154                         (*Response).StatusCode\ngithub.com/valyala/fasthttp@v1.43.0/server.go:2338                      (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.43.0/workerpool.go:224                   (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.43.0/workerpool.go:196                   (*workerPool).getCh.func1\nruntime/asm_amd64.s:1594                                                goexit"

A timeout situation.

However in the chrome browser console I get a 500 in the attempt to call http://authelia:9091/api/firstfactor
and in
docker exec -it nginx curl http://authelia:9091/api/firstfactor
i get  405 Method not allowed

**My question is how do I troubleshoot this - where do I go from here?**

### My docker-compose.yml for nginx:

version: '3.3'

networks:
  auth_net:
    external: true

services:
 authelia:
    image: authelia/authelia:latest
    container_name: authelia
    volumes:
      - /opt/nginx/authelia/config:/config
      - /opt/nginx/hosts:/etc/hosts:ro
    ports:
      - "9091:9091"
      - "9093:9093"
    restart: unless-stopped
    healthcheck:
      disable: true
    environment:
      - TZ=America/Denver
    depends_on:
      - redis
    networks:
      auth_net:
        ipv4_address: 192.168.1.11

 redis:
    image: redis:alpine
    container_name: redis-container
    user: "1000:1000"
    ports:
      - 6379:6379
    volumes:
      - ./docker/data/redis:/data
    restart: unless-stopped
    environment:
      - TZ=America/Denver
    networks:
      auth_net:
        ipv4_address: 192.168.1.12

 nginx: 
    image: nginx:latest
    container_name: nginx
    ports:
       - "80:80"
       - "443:443"
    volumes:
       - /opt/nginx/config:/etc/nginx
       - /opt/nginx/authelia/certs:/etc/authelia
       - /opt/nginx/certs:/etc/ssl
       - /opt/nginx/web_content:/usr/share/nginx/html:ro
    networks:
      auth_net:
        ipv4_address: 192.168.1.10
    depends_on:
      - authelia

### my nginx configuration file:

events { }

http {
  server {
      listen 80;
      server_name twofactor.jnorris.org;
      return 301 https://$server_name$request_uri;
  }

  server {
      listen 443 ssl;

      server_name twofactor.jnorris.org;

      ssl_certificate /etc/ssl/server.pem;
      ssl_certificate_key  /etc/ssl/server.key;

      location / {
          proxy_pass http://authelia:9091;
          proxy_set_header Host $host;
          proxy_set_header X-Real-IP $remote_addr;
          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header X-Forwarded-Proto $scheme;

          auth_request /authelia/api/verify?on_error=auth_redirect;
          error_page 401 = @auth_redirect;
      }

      location @auth_redirect {
          return 302 /authelia/signin?rd=https://$host$request_uri;
      }

      location /authelia {
          proxy_pass http://authelia:9091;
          proxy_set_header Host $host;
          proxy_set_header X-Real-IP $remote_addr;
          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header X-Forwarded-Proto $scheme;
   	  proxy_read_timeout 60s;
      }

      location = /favicon.ico {
          access_log off;
          log_not_found off;
      }

      location = /index.html {
          root /usr/share/nginx/html;
          index index.html;
      }
  }
  
}

### my authelia configuration file:

server:
  host: 0.0.0.0 
  port: 9091
  buffers:
    read: 4096
    write: 4096
  timeouts:
    read: 6s
    write: 6s
    idle: 30s

log:
  level: debug 

jwt_secret: blah

totp:
  issuer: authelia.com
  disable: false
  algorithm: sha1
  period: 30
  digits: 6
  skew: 1
  secret_size: 32

session:
  name: authelia_session
  secret: blah
  expiration: 3600 # 1 hour
  inactivity: 7200 # 2 hours
  domain: jnorris.org

  redis:
    host: 192.168.1.12
    port: 6379

storage:
   encryption_key: blah
   local:
     path: /config/db.sqlite3

notifier:
  filesystem:
    filename:  /config/contact.txt

authentication_backend:
  refresh_interval: 5m
  password_reset:
    disable: false
    custom_url: ""

  file:
    path: /config/users_database.yml
    password:
      algorithm: argon2id
      iterations: 1
      salt_length: 16
      parallelism: 8
      memory: 64

access_control:
  default_policy: deny
  rules:
    - domain: "auth.jnorris.org"
      policy: bypass
    - domain: "twofactor.jnorris.org"
      policy: two_factor

**Comment by james-d-elliott** [comment]
Where did you get the nginx config from? Seems kind of broken if someone is suggesting this configuration but I may be missing something. You're also missing some of the required headers I think: https://www.authelia.com/integration/proxies/introduction/#required-headers

**Reply by jozefn** [reply]
Thanks for your observations James - I went back and started fresh. Got a clean version of nginx and made sure that I could use the proxy to redirect to my test websites. All went well with this:
server {
    listen       80;
    listen  [::]:80;

    server_name  localhost;

    #access_log  /var/log/nginx/host.access.log  main;

    location / {
        root   /usr/share/nginx/html;
        index  index.html index.htm;
    }

    location = /onefactor/ {
       rewrite ^/onefactor(.*)$ /$1 break;
       proxy_pass http://onefactor.jnorris.net;
    }

    location = /twofactor/ {
       rewrite ^/twofactor(.*)$ /$1 break;
       proxy_pass  http://twofactor.jnorris.net;
    }


    #error_page  404              /404.html;

    # redirect server error pages to the static page /50x.html
    #
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

}

so with the above https://localhost/onefactor - takes me to my http://onefactor.jnorris.net

So I made some adjustments with the following:

events {}

http {

  server {
  
    listen       80;
    listen  [::]:80;
    listen  443 ssl;
   
    server_name  localhost;
  
    location / {
      root   /usr/share/nginx/html;
      index  index.html index.htm;
    }
   
    include /etc/nginx/snippets/ssl.conf;
  
    set $upstream http://authelia:9091;
  
    location = /auth/ {
      rewrite ^/auth(.*)$ /$1 break;
      proxy_pass http://auth.jnorris.net;
    }
  
  
    location = /onefactor/ {
      rewrite ^/onefactor(.*)$ /$1 break;
      include /etc/nginx/snippets/proxy.conf;
      include /etc/nginx/snippets/authelia-authrequest.conf;
      proxy_pass $upstream;
    }
    
    location = /twofactor/ {
      rewrite ^/twofactor(.*)$ /$1 break;
      include /etc/nginx/snippets/proxy.conf;
      include /etc/nginx/snippets/authelia-authrequest.conf;
      proxy_pass $upstream;
  
    }
    
    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
      root   /usr/share/nginx/html;
    }
  
  } 

}


when I do the https://localhost/onefactor I get 

The page you are looking for is temporarily unavailable. Please try again later.
I used only the snippets from the site and followed the configuration.  

The logs are clean and I am not sure how to troubleshoot this issue.

Thanks for your patience

**Reply by james-d-elliott** [reply]
Yeah no stress. I'll do everything I can to help. For reference since I forgot to mention this before we return 405 instead of 404 if the endpoint exists but doesn't support the particular method verb (GET vs POST etc).

For the sake of testing can you change your log settings to have level trace?

```yaml
log:
  level: trace
```

Also can you check the nginx logs and authelia logs again after this? Also why are you configuring those rewrites? I feel like they may not be required and are potentially an element of the issue or will lead to future issues.

**Reply by jozefn** [reply]
I had the rewrites in because nginx was sending me to onefactor.jnorris.net/onefactor and there is not onefactor path.  I will put trace and send logs.  Thank you so much!  this is pretty perplexing. First time with authelia and nginx - been mainly in apache

**Reply by jozefn** [reply]
[josephnorris@localhost config]$ sudo docker logs authelia
time="2023-08-07T18:08:37-06:00" level=info msg="Authelia v4.37.5 is starting"
time="2023-08-07T18:08:37-06:00" level=info msg="Log severity set to trace"
time="2023-08-07T18:08:37-06:00" level=trace msg="Starting scan of directory  for certificates"
time="2023-08-07T18:08:37-06:00" level=trace msg="Finished scan of directory  for certificates"
time="2023-08-07T18:08:37-06:00" level=info msg="Storage schema is being checked for updates"
time="2023-08-07T18:08:37-06:00" level=info msg="Storage schema is already up to date"
time="2023-08-07T18:08:37-06:00" level=info msg="Initializing server for non-TLS connections on '[::]:9091' path '/'"

**Reply by jozefn** [reply]
I also removed the rewrite statements same thing happening:

The page you are looking for is temporarily unavailable. Please try again later.
Website Administrator
Something has triggered missing webpage on your website. This is the default error page for nginx that is distributed with Rocky Linux. It is located /usr/share/nginx/html/50x.html

You should customize this error page for your own site or edit the error_page directive in the nginx configuration file /etc/nginx/nginx.conf.

For information on Rocky Linux, please visit the Rocky Linux website. The documentation for Rocky Linux is available on the Rocky Linux website.

**Reply by james-d-elliott** [reply]
Ah yep, nginx error logs should tell you what's going on. I suspect it'll be an issue with the IP etc of the proxy pass.

**Reply by jozefn** [reply]
but the above was all I got from setting level to trace and I understand that I should have gotten a lot more.

here is the messages file from host:
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: [3613:3638:0807/184014.730630:ERROR:cert_verify_proc_builtin.cc(701)] CertVerifyProcBuiltin for localhost failed:
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: ----- Certificate i=0 (CN=demo.mlopshub.com,OU=MlopsHub Dev,O=MLopsHub,L=San Fransisco,ST=California,C=US) -----
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: ERROR: No matching issuer found
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: [3729:3746:0807/184014.732189:ERROR:ssl_client_socket_impl.cc(980)] handshake failed; returned -1, SSL error code 1, net_error -202
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: [3729:3746:0807/184014.732880:ERROR:ssl_client_socket_impl.cc(980)] handshake failed; returned -1, SSL error code 1, net_error -202
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: [3729:3746:0807/184014.765818:ERROR:ssl_client_socket_impl.cc(980)] handshake failed; returned -1, SSL error code 1, net_error -202
Aug  7 18:40:14 localhost google-chrome.desktop[3619]: [3729:3746:0807/184014.767028:ERROR:ssl_client_socket_impl.cc(980)] handshake failed; returned -1, SSL error code 1, net_error -202
Aug  7 18:40:14 localhost NetworkManager[995]: <info>  [1691455214.9428] policy: set-hostname: set hostname to 'localhost.localdomain' (no hostname found)
Aug  7 18:40:17 localhost NetworkManager[995]: <info>  [1691455217.9126] policy: set-hostname: set hostname to 'localhost.localdomain' (no hostname found)
Aug  7 18:40:20 localhost NetworkManager[995]: <info>  [1691455220.8822] policy: set-hostname: set hostname to 'localhost.localdomain' (no hostname found)
Aug  7 18:40:22 localhost systemd[1]: Starting dnf makecache...
Aug  7 18:40:23 localhost dnf[10744]: Docker CE Stable - x86_64                        14 kB/s | 3.5 kB     00:00

**Reply by jozefn** [reply]
also the log files for nginx are as follows:
more /var/lib/docker/containers/7224574f9d7e39c06feee5f8ff5ec3a7b727868344a2963dc28580db55d0927f/7224574f9d7e39c06feee5f8ff5ec3a7b727868344a2963dc28580db55d0927f-json.log
{"log":"/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration\n","stre
am":"stdout","time":"2023-08-08T00:39:49.750947348Z"}
{"log":"/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/\n","stream":"stdout","time":"20
23-08-08T00:39:49.751015413Z"}
{"log":"/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh\n","stream":"stdout
","time":"2023-08-08T00:39:49.752208639Z"}
{"log":"10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf is not a file or does not exist\n",
"stream":"stdout","time":"2023-08-08T00:39:49.753816847Z"}
{"log":"/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh\n","stream":"stdout","time"
:"2023-08-08T00:39:49.754516103Z"}
{"log":"/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh\n","stream":"stdout","t
ime":"2023-08-08T00:39:49.75455073Z"}
{"log":"/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh\n","stream":"stdout","t
ime":"2023-08-08T00:39:49.757606893Z"}
{"log":"/docker-entrypoint.sh: Configuration complete; ready for start up\n","stream":"stdout","time":"2023-08-08
T00:39:49.759327607Z"}
{"log":"2023/08/08 00:40:14 [error] 20#20: *3 no resolver defined to resolve authelia, client: 192.168.1.1, serve
r: localhost, request: \"GET /onefactor/ HTTP/1.1\", subrequest: \"/authelia\", host: \"localhost\"\n","stream":"
stderr","time":"2023-08-08T00:40:14.739331165Z"}
{"log":"2023/08/08 00:40:14 [error] 20#20: *3 auth request unexpected status: 502 while sending response to clien
t, client: 192.168.1.1, server: localhost, request: \"GET /onefactor/ HTTP/1.1\", host: \"localhost\"\n","stream"
:"stderr","time":"2023-08-08T00:40:14.739393734Z"}
{"log":"192.168.1.1 - - [08/Aug/2023:00:40:14 +0000] \"GET /onefactor/ HTTP/1.1\" 500 3404 \"-\" \"Mozilla/5.0 (X
11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36\"\n","stream":"stdout","t
ime":"2023-08-08T00:40:14.73942553Z"}
{"log":"192.168.1.1 - - [08/Aug/2023:00:40:14 +0000] \"GET /onefactor/nginx-logo.png HTTP/1.1\" 404 555 \"https:/
/localhost/onefactor/\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0
 Safari/537.36\"\n","stream":"stdout","time":"2023-08-08T00:40:14.781112869Z"}
{"log":"2023/08/08 00:40:14 [error] 20#20: *6 open() \"/usr/share/nginx/html/onefactor/nginx-logo.png\" failed (2
: No such file or directory), client: 192.168.1.1, server: localhost, request: \"GET /onefactor/nginx-logo.png HT
TP/1.1\", host: \"localhost\", referrer: \"https://localhost/onefactor/\"\n","stream":"stderr","time":"2023-08-08
T00:40:14.781184509Z"}
{"log":"2023/08/08 00:40:14 [error] 20#20: *7 open() \"/usr/share/nginx/html/onefactor/poweredby.png\" failed (2:
 No such file or directory), client: 192.168.1.1, server: localhost, request: \"GET /onefactor/poweredby.png HTTP
/1.1\", host: \"localhost\", referrer: \"https://localhost/onefactor/\"\n","stream":"stderr","time":"2023-08-08T0
0:40:14.784801426Z"}
{"log":"192.168.1.1 - - [08/Aug/2023:00:40:14 +0000] \"GET /onefactor/poweredby.png HTTP/1.1\" 404 555 \"https://
localhost/onefactor/\" \"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 
Safari/537.36\"\n","stream":"stdout","time":"2023-08-08T00:40:14.784934745Z"}

**Reply by jozefn** [reply]
Thanks for your help on this - I really want to  figure it out.

**Reply by james-d-elliott** [reply]
No worries. I can't really see  what the issue is, but it looks kind of like it may be a misunderstanding from moving to nginx but we can probably help either way. We generally recommend running authelia and protected applications as subdomains of a particular domain. For example `https://auth.example.com`, `https://app.example.com`, `https://another.example.com`, etc.  Is this the style you want to implement or are you trying to do subpath? For example `https://example.com/auth`, `https://example.com/app`, `https://example.com/another`.

**Reply by jozefn** [reply]
ok so then if I create a subdomain that is under my factious domain jnorris.org that is resolved via hosts file - this might work?

**Comment by jozefn** [comment]
yes what you describe above is what I am after. https://twofactor.jnorris.org https://onefactor.jnorris.org  - just not sure how to implement this.

**Reply by james-d-elliott** [reply]
Yep no worries, I'll try to put together an example today which may be able to communicate how this is achievable with nginx incorporating one example of authelia itself, one which integrates with authelia, and one which does not.

**Comment by james-d-elliott** [comment]
Each file is commented. Replace `http://authelia:9091` with your specific relevant internal endpoint for authelia, and  `auth.example.com` with your domain. Should show you how to do subdomains in an organized way that's  easy to configure. Would note the image I used for nginx has a module described here (the set_misc module): https://authelia-next.netlify.app/integration/proxies/nginx/#requirements

```yaml
# file: ./docker-compose.yml
# description: example docker compose to go with this example. I'd recommend using the image we've used here as it has an semi-important module.

services:
  nginx:
    container_name: nginx
    image: lscr.io/linuxserver/nginx:latest
    restart: unless-stopped
    networks:
      net:
        aliases:
          - "nginx.internal"
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - ${PWD}/data/nginx/snippets:/config/nginx/snippets:ro
      - ${PWD}/data/certificates:/config/nginx/certificates:ro
      - ${PWD}/data/nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ${PWD}/data/nginx/sites-enabled.d:/config/nginx/sites-enabled.d
networks:
  net:
    external: true
    name: net
```

```nginx
# file: ./data/nginx/nginx.conf
# description: the nginx main config file.

#include /etc/nginx/modules/*.conf;

user  nginx;
worker_processes  auto;

error_log  /var/log/nginx/error.log notice;
pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    keepalive_timeout  65;

    #gzip  on;

    # includes all of the other configs so we can have virtual servers.
    include /etc/nginx/sites-enabled.d/*.conf;
}
```

```nginx
# file: ./data/nginx/sites-enabled.d/default.conf
# description: default fallback server for any other request that isn't explicitly matched. Prevents strange issues.

server {
    listen 80 default_server;
    listen [::]:80 default_server;

    server_name _;

    location / {
        return 301 https://$host$request_uri;
    }
}

server {
    listen 443 ssl http2 default_server;
    listen [::]:443 ssl http2 default_server;

    server_name _;

    include /config/nginx/snippets/ssl.conf;

    location / {
        root   /usr/share/nginx/html;
        index  index.html index.htm;
    }

    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }
}
```

```nginx
# file: ./data/nginx/sites-enabled.d/auth.conf
# description: authelia portal

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    # this directs nginx to respond on this virtual server for any subdomain starting wth "auth.".
    server_name auth.*;

    include /config/nginx/ssl.conf;

    # change this to the correct info for the authelia backend
    set $upstream http://authelia:9091;

    location / {
        include /config/nginx/proxy.conf;

        proxy_pass $upstream;
    }

    location /api/verify {
        proxy_pass $upstream;
    }
}
```

```nginx
# file: ./data/nginx/sites-enabled.d/app.conf
# description: example application WITH authelia.

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    # this directs nginx to respond on this virtual server for any subdomain starting wth "app.".
    server_name app.*;

    # example ssl config.
    include /config/nginx/snippets/ssl.conf;
    
    # enable authelia virtual location.
    include /config/nginx/snippets/authelia-location.conf;

    # change this to the correct info for the app backend
    set $upstream http://app:8080;

    location / {
        include /config/nginx/snippets/proxy.conf;
        include /config/nginx/snippets/authelia-authrequest.conf;

        proxy_pass $upstream;
    }
}
```

```nginx
# file: ./data/nginx/sites-enabled.d/x.conf
# description: example application WITHOUT authelia.

server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    # this directs nginx to respond on this virtual server for any subdomain starting wth "x.".
    server_name x.*;

    # example ssl config.
    include /config/nginx/snippets/ssl.conf;

    # change this to the correct info for the app backend
    set $upstream http://app:8081;

    location / {
        include /config/nginx/snippets/proxy.conf;

        proxy_pass $upstream;
    }
}
```

```nginx
# file: ./data/nginx/snippets/ssl.conf
# description: example ssl.conf file with custom certs. If you're using LE or something this may require some work.

ssl_certificate /config/nginx/certificates/wildcard.pem;
ssl_certificate_key /config/nginx/certificates/wildcard.key;
ssl_session_timeout 1d;
ssl_session_cache shared:MozSSL:10m;  # about 40000 sessions
ssl_session_tickets off;

# modern configuration
ssl_protocols TLSv1.3;
ssl_prefer_server_ciphers off;

# replace with the IP address of your resolver
resolver 127.0.0.11;
```

```nginx
# file: ./data/nginx/snippets/authelia-location.conf
# description: special internal / virtual location to perform the authz request. The first commented line is for a future Authelia version. This should exist in the same area on each enabled sever that you want authelia on.

#set $upstream_authelia http://authelia:9091/api/authz/auth-request;
set $upstream_authelia http://authelia:9091/api/verify;

## Virtual endpoint created by nginx to forward auth requests.
location /authelia {
    ## Essential Proxy Configuration
    internal;
    proxy_pass $upstream_authelia;

    ## Headers
    ## The headers starting with X-* are required.
    proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
    proxy_set_header X-Original-Method $request_method;
    proxy_set_header X-Forwarded-Method $request_method;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header Content-Length "";
    proxy_set_header Connection "";

    ## Basic Proxy Configuration
    proxy_pass_request_body off;
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; # Timeout if the real server is dead
    proxy_redirect http:// $scheme://;
    proxy_http_version 1.1;
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 4 32k;
    client_body_buffer_size 128k;

    ## Advanced Proxy Configuration
    send_timeout 5m;
    proxy_read_timeout 240;
    proxy_send_timeout 240;
    proxy_connect_timeout 240;
}
```

```nginx
# file: ./data/nginx/snippets/authelia-authrequest.conf
# description: special directives to make nginx leverage the authelia-location.conf to make the authz request. Any location these are in will try to utilize that endpoint. This should exist in the same area on each enabled sever that you want authelia on.

## Send a subrequest to Authelia to verify if the user is authenticated and has permission to access the resource.
auth_request /authelia;

## Set the $target_url variable based on the original request.
#auth_request_set $target_url $scheme://$http_host$request_uri;
set_escape_uri $target_url $scheme://$http_host$request_uri;

## Save the upstream response headers from Authelia to variables.
auth_request_set $user $upstream_http_remote_user;
auth_request_set $groups $upstream_http_remote_groups;
auth_request_set $name $upstream_http_remote_name;
auth_request_set $email $upstream_http_remote_email;
# auth_request_set $redirection $upstream_http_location;

## Inject the response headers from the variables into the request made to the backend.
proxy_set_header Remote-User $user;
proxy_set_header Remote-Groups $groups;
proxy_set_header Remote-Name $name;
proxy_set_header Remote-Email $email;

## If the subreqest returns 200 pass to the backend, if the subrequest returns 401 redirect to the portal.
error_page 401 =302 https://auth.example.com/?rd=$target_url;
#error_page 401 =302 $redirection;
```

```nginx
# file: ./data/nginx/snippets/proxy.conf
# description: meta file for use with basically every reverse proxied host which sets special headers, some of these are essential for authelia specifically.

## Headers
proxy_set_header Host $host;
proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-URI $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_set_header X-Forwarded-For $remote_addr;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $connection_upgrade;

## Basic Proxy Configuration
client_body_buffer_size 128k;
proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; ## Timeout if the real server is dead.
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 64 256k;

## Trusted Proxies Configuration
## Please read the following documentation before configuring this:
##     https://www.authelia.com/integration/proxies/nginx/#trusted-proxies
# set_real_ip_from 10.0.0.0/8;
# set_real_ip_from 172.16.0.0/12;
# set_real_ip_from 192.168.0.0/16;
# set_real_ip_from fc00::/7;
real_ip_header X-Forwarded-For;
real_ip_recursive on;

## Advanced Proxy Configuration
send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;
```

**Reply by jozefn** [reply]
Thank you - I will give this a shot.

**Comment by jozefn** [comment]
Hi can you explain a bit more about this parameter:
# replace with the IP address of your resolver
resolver 127.0.0.11;

in the context of what I am trying to do with the hosts file and 127.0.0.1?

My hosts file is like this:
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4 
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6

192.168.1.10 auth.jnorris.org auth
192.168.1.11 authelia.jnorris.org authelia

**Reply by james-d-elliott** [reply]
The 127.0.0.11 IP is the IP of the docker DNS server, it should always be a DNS server I think from memory. It's used to resolve things like `http://authelia:9091` instead of using an IP.

**Reply by james-d-elliott** [reply]
Side note if all of your containers are on docker and on the same network it should just work as  is.

**Reply by jozefn** [reply]
James - thank you for your help.  I put this set up into place and did a docker compose but I only see the nginx container but not the authelia container. Do I add authelia to the docker-compose.yml?

**Reply by james-d-elliott** [reply]
Yeah you can include authelia as per normal and join it to the same docker network and it should all work. If you do exactly the same containers as you previously did (and same network / network name / network config) it should work. My example was intended to be adapted to your particular needs.

**Comment by jozefn** [comment]
Sorry James - I really don't know what I am doing wrong.  Can I make you collaborator on so you can look at my repository?

**Reply by james-d-elliott** [reply]
Sure, go ahead and I'll take a look.

**Reply by james-d-elliott** [reply]
You're also welcome to contact me over Discord or Matrix if you want by the way (see my profile).

**Comment by jozefn** [comment]
*github

**Reply by jozefn** [reply]
I am on Discord - jozefn999 and just sent you an invite.
